storage: don't send log truncations through Raft

tbg opened the pull request:
Please discuss - this looks like an easy win. I just whipped this up on a whim,
so I hope I'm not missing something obvious here.

--

Sending log truncations through Raft is inefficient: the Raft log is not itself
part of the replicated state. Instead, we only replicate the TruncatedState and,
as a side effect, ClearRange() the affected key range.

This is an individual performance optimization whose impact we should measure;
anecdotally it always looked like we were doing a lot of work for truncations
during a write-heavy workload; this should alleviate this somewhat).

It also removes one migration concern for #16809, see
https://github.com/cockroachdb/cockroach/pull/16809#issuecomment-314533973.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16993)
<!-- Reviewable:end -->

petermattis commented:
Doesn't this add a migration concern? In a mixed-version cluster, a server with this PR will not generate a Raft command that truncates the Raft log for a 1.0.x server.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/replica_proposal.go, line 762 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-Kon3ZQaspxHmh3ivdJW:-Kon3ZQaspxHmh3ivdJX:b-ao090v) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6588864c0b879cce55ebec158579dc5300af8c/pkg/storage/replica_proposal.go#L762)):*
> ```Go
> 				keys.RaftLogKey(r.RangeID, newTruncState.Index).PrefixEnd(),
> 			)
> 			if err := r.store.engine.ClearRange(start, end); err != nil {
> ```

We definitely don't want to use `ClearRange` here as this would create an unreasonable number of range deletion tombstones.  Use `ClearIterRange` instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-Kon3pjtrVSzZxRs_CKN:by8m4i1)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> Doesn't this add a migration concern?

Does it? The below-Raft side effect is optional, and above Raft we can do what we want. The only thing that comes to mind is that when you run a mixed cluster for a long time, replicas running the old version may not actually remove their log entries.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-Kon4MTJhZRtu2O9UhK0:b-uo7gg2)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> The only thing that comes to mind is that when you run a mixed cluster for a long time, replicas running the old version may not actually remove their log entries.

That was my concern. That seems like it could be problematic. Maybe I'm not thinking it through properly. 

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-Kon4qe4MN_rYMf6DbBW:bj7z8a1)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Could be problematic, but my point is rather that it's easy to migrate this change - the "new version" just has to emit old-style truncations as long as it thinks there are still old nodes around. Pretty simple compared to other migrations on the table right now, and should be easy to cover this.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/replica_proposal.go, line 762 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-Kon3ZQaspxHmh3ivdJW:-Kon5XC2co9wzetuJ-5x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6588864c0b879cce55ebec158579dc5300af8c/pkg/storage/replica_proposal.go#L762)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We definitely don't want to use `ClearRange` here as this would create an unreasonable number of range deletion tombstones.  Use `ClearIterRange` instead.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-Kon5JBgrhzqAeOTZGnM:b2whe90)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_command.go, line 1908 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KonHdMNxCDbgordNXD_:-KonHdMNxCDbgordNXDa:bjc2nza) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fac16a368d434c67cb9345cbfe583c2747913bd/pkg/storage/replica_command.go#L1908)):*
> ```Go
> 	ms, err := iter.ComputeStats(start, end, 0 /* nowNanos */)
> 	if err != nil {
> 		return EvalResult{}, errors.Wrap(err, "while computing stats of Raft log freed by truncation")
> ```

"could not compute ..." is more in line with our convention, I think.

---

*[pkg/storage/replica_proposal.go, line 762 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-Kon3ZQaspxHmh3ivdJW:-KonGx1eB0NHtise3MCl:bmvdrxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6588864c0b879cce55ebec158579dc5300af8c/pkg/storage/replica_proposal.go#L762)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Would be good to add a comment explaining why this doesn't `ClearRange`, which naively seems better.

---

*[pkg/storage/replica_proposal.go, line 764 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KonH50H-2404dMHLNMA:-KonH50H-2404dMHLNMB:b6u04xf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fac16a368d434c67cb9345cbfe583c2747913bd/pkg/storage/replica_proposal.go#L764)):*
> ```Go
> 			iter := r.store.engine.NewIterator(false /* !prefix */)
> 			if err := r.store.engine.ClearIterRange(iter, start, end); err != nil {
> 				iter.Close()
> ```

extract below, or wrap in a function (instead of the scope) and defer it

---

*[pkg/storage/replica_proposal.go, line 776 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KonH81NYlFnzOjMBrBO:-KonH81OBmWYFh4PcwsM:b-bwx527) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fac16a368d434c67cb9345cbfe583c2747913bd/pkg/storage/replica_proposal.go#L776)):*
> ```Go
> 
> 		// Truncate the sideloaded storage.
> 		{
> ```

this isn't a real scope, right? could do without it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_command.go, line 1908 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KonHdMNxCDbgordNXD_:-KonO3mo6H4S2lu5tXEV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fac16a368d434c67cb9345cbfe583c2747913bd/pkg/storage/replica_command.go#L1908)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

"could not compute ..." is more in line with our convention, I think.
</blockquote></details>

Done.

---

*[pkg/storage/replica_proposal.go, line 762 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-Kon3ZQaspxHmh3ivdJW:-KonOL5bwfcoIQnJevac:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6588864c0b879cce55ebec158579dc5300af8c/pkg/storage/replica_proposal.go#L762)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Would be good to add a comment explaining why this doesn't `ClearRange`, which naively seems better.
</blockquote></details>

Done.

---

*[pkg/storage/replica_proposal.go, line 764 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KonH50H-2404dMHLNMA:-KonOQv93u5qAupjxSdV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fac16a368d434c67cb9345cbfe583c2747913bd/pkg/storage/replica_proposal.go#L764)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

extract below, or wrap in a function (instead of the scope) and defer it
</blockquote></details>

Done.

---

*[pkg/storage/replica_proposal.go, line 776 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KonH81NYlFnzOjMBrBO:-KonOSAR_PRbMWXaQDY1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fac16a368d434c67cb9345cbfe583c2747913bd/pkg/storage/replica_proposal.go#L776)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this isn't a real scope, right? could do without it.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_proposal.go, line 762 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-Kon3ZQaspxHmh3ivdJW:-KonPBImw8z3QPR1Tq6E:bc5kud9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6588864c0b879cce55ebec158579dc5300af8c/pkg/storage/replica_proposal.go#L762)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

(Should have clarified: I added a TODO for Peter to add a comment...)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_proposal.go, line 764 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KonQYXTGgLngXwWWuNH:-KonQYXTGgLngXwWWuNI:b-dcbmya) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2370573f87d6d325077665b0b5ad9e774a0e658/pkg/storage/replica_proposal.go#L764)):*
> ```Go
> 			iter := r.store.engine.NewIterator(false /* !prefix */)
> 			// TODO(petermattis): add a comment explaining why using ClearRange() below
> 			// is ill-advised.
> ```

Range deletion tombstones add a per-query overhead. The RocksDB folks recommend limiting the number that are used. Currently we only use them when deleting an entire range. Add such tombstones for Raft log truncation is likely to result in an excessive number. I should really investigate how the range deletion tombstones are indexed within sstables to understand this better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_proposal.go, line 764 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KonQYXTGgLngXwWWuNH:-KonRZVQNuVjLB3zmwrm:b1w9vyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2370573f87d6d325077665b0b5ad9e774a0e658/pkg/storage/replica_proposal.go#L764)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Range deletion tombstones add a per-query overhead. The RocksDB folks recommend limiting the number that are used. Currently we only use them when deleting an entire range. Add such tombstones for Raft log truncation is likely to result in an excessive number. I should really investigate how the range deletion tombstones are indexed within sstables to understand this better.
</blockquote></details>

Interested in understanding this better, too. The tombstones we make here are essentially `[0,i` for increasing `i`, so it seems that any that hit the same level/sstable/whatever should, in theory, coagulate and leave only one (also seeing that we never "break them up" by writing over them).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
In addition to Peter's concern about migrating to this change from 1.0, this change moves logic downstream of raft. This will generally make *future* migrations harder. Maybe it's justifiable in this case, but we'll need to see what the performance numbers are like (or maybe we want to do this for other reasons, like the way it would interact with the migration needed for @irfansharif's change). 

---

Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-KornkvIKoSnDbo2o0nx:be3t78o)*
<!-- Sent from Reviewable.io -->

petermattis commented:
One nice bit this change would enable is we could defer the actual deletion of Raft log entries in order to avoid needing to sync the normal (non-Raft log) engine. 

---

Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-KorrvUzSEyaAMQZwLoi:bfra9sc)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I'll play with some performance tests. We'll see what happens; the original
motivation for whipping this up was to facilitate Irfan's migration. Will
update, probably next week.

On Wed, Jul 12, 2017 at 1:38 PM Peter Mattis <notifications@github.com>
wrote:

> One nice bit this change would enable is we could defer the actual
> deletion of Raft log entries in order to avoid needing to sync the normal
> (non-Raft log) engine.
> ------------------------------
>
> Review status: 1 of 2 files reviewed at latest revision, 2 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-KorrvUzSEyaAMQZwLoi:bfra9sc>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/16993#issuecomment-314842564>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135GI-3qaCCEwzezZS0POjcwAN8A-8ks5sNQSmgaJpZM4OUrll>
> .
>
-- 

-- Tobias

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_proposal.go, line 764 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KonQYXTGgLngXwWWuNH:-KoseiOc3DXV2u8jFJZS:b-5vfj1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2370573f87d6d325077665b0b5ad9e774a0e658/pkg/storage/replica_proposal.go#L764)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Interested in understanding this better, too. The tombstones we make here are essentially `[0,i` for increasing `i`, so it seems that any that hit the same level/sstable/whatever should, in theory, coagulate and leave only one (also seeing that we never "break them up" by writing over them).
</blockquote></details>

Range tombstones are stored in a per-sstable "meta" block that is loaded when the sstable is open. I'm not seeing any code which merges adjacent range tombstones together, though I did run across code which "collapses deletions". That could be it. 

When retrieving a key, all of the tombstones contained in each sstable encountered during retrieval are added to a `RangeDelAggregator` which then answers whether the key has been deleted. It looks like some effort has been put into making `RangeDelAggregator` efficient, but note that all of the tombstones for a given sstable are added to it even when those tombstones do not overlap the key being retrieved or the span being iterated. I imagine this is where the warning about not using too many tombstones comes from.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on pkg/storage/replica_proposal.go:
```diff
@@ -737,17 +737,49 @@ func (r *Replica) handleReplicatedEvalResult(
+				keys.RaftLogKey(r.RangeID, newTruncState.Index).PrefixEnd(),
+			)
+			iter := r.store.engine.NewIterator(false /* !prefix */)
+			if err := r.store.engine.ClearIterRange(iter, start, end); err != nil {
```
This generates per-key tombstones. Why not use `ClearRange` instead?

spencerkimball replied:
nm...i just read the discussion.

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_proposal.go, line 764 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KonQYXTGgLngXwWWuNH:-Kp1D14D-OhJYhuVUO8G:bqco7on) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2370573f87d6d325077665b0b5ad9e774a0e658/pkg/storage/replica_proposal.go#L764)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Range tombstones are stored in a per-sstable "meta" block that is loaded when the sstable is open. I'm not seeing any code which merges adjacent range tombstones together, though I did run across code which "collapses deletions". That could be it. 

When retrieving a key, all of the tombstones contained in each sstable encountered during retrieval are added to a `RangeDelAggregator` which then answers whether the key has been deleted. It looks like some effort has been put into making `RangeDelAggregator` efficient, but note that all of the tombstones for a given sstable are added to it even when those tombstones do not overlap the key being retrieved or the span being iterated. I imagine this is where the warning about not using too many tombstones comes from.
</blockquote></details>

As an additional datapoint for future archaeologists, SCATTER at one point created 1000s of these tombstones, and it really got RocksDB down to its knees: https://github.com/cockroachdb/cockroach/pull/16249#issuecomment-314920699

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ok, I take that back. I don't actually want to play with performance tests just yet. Instead, I first want to discuss the merits of this PR under the assumption that performance stays pretty much the same for now. The relevant points here:

- we'll need to migrate this. It's straightforward with the in-flight PR #16977
- we're moving logic downstream of Raft. However, we can easily migrate it upstream again, without a real migration, though I don't think that's going to happen.
- the big upshot is hopefully a large reduction in complexity for @irfansharif's PR: log truncation is one of the odd cases that requires a RaftWriteBatch. #16749 is the only other one, and there the (correct) solution also involves going downstream of Raft for a Raft-related write. So, after solving both of those, I *think* `RaftWriteBatch` can go? cc @irfansharif
- as @petermattis pointed out, after @irfansharif's change, we should be able to not sync the base engine on truncation changes but do it only as we actually clear the log entries (which can be delayed as we see fit). So for 1000 log truncations  across many ranges, we'll only have to sync once if that's how we set it up.

For performance, I think it'll be more obvious that it's a pro when the last bullet comes in. Right now, as the change is written, we save only a few on-the-wire bytes. That's nice, but not really something to write home about, and it's hard to demonstrate it, too.

My suggestion is the following:

1. leave this change open until we have the migration framework in place, at which point I adapt & merge it
1. @irfansharif rebases his change on top of this (and #16749 which I'll prioritize)
1. once his change lands, we optimize the truncation sync as outlined above (punting to 1.2 if we really run out of time)

@bdarnell @petermattis Thoughts?

@irfansharif please confirm that the above plan would really simplify your PR as I imagine.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-KpG7DIiSs1k-xNqKZP1:b-m6orze)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
SGTM

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_proposal.go, line 762 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-Kon3ZQaspxHmh3ivdJW:-KpGR9RKcYZcV2sMNT1R:b-gma5bk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6588864c0b879cce55ebec158579dc5300af8c/pkg/storage/replica_proposal.go#L762)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

(Should have clarified: I added a TODO for Peter to add a comment...)
</blockquote></details>

Looks like you haven't pushed that TODO. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-KpGRjBp8vV5araWSFgl:bb73djm)*
<!-- Sent from Reviewable.io -->

irfansharif commented:
As an aside re: delaying _actual_ log entry deletions, this is what I was alluding to [here (comment)](https://github.com/cockroachdb/cockroach/issues/16948#issuecomment-314208214).

>  So, after solving both of those, I think RaftWriteBatch can go? 

Yup (`WriteBatch.RaftData` you mean).

> So for 1000 log truncations across many ranges, we'll only have to sync once if that's how we set it up.

Yup, this batching and lazy _actual_ deletion will be crucial. I'll be posting to #16624 shortly with some recent numbers but log truncations (and by extension, base engine syncs) happen far too often right now to bring about desired speedups.

> @irfansharif please confirm that the above plan would really simplify your PR as I imagine.

Yup.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-KpGfmwYy1MdmW6tOLgr:b-t91knk)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Cool. PR for https://github.com/cockroachdb/cockroach/issues/16749 is about to be posted, too.

petermattis commented:
SGTM too

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-KpGlbgjDlx4JnysGyUz:bv26eti)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
PTAL: I rebased on top of https://github.com/cockroachdb/cockroach/pull/17068.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/replica_proposal.go, line 762 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-Kon3ZQaspxHmh3ivdJW:-KpHNl-7jEFoMBRV2Qnx:b-f8sf1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a6588864c0b879cce55ebec158579dc5300af8c/pkg/storage/replica_proposal.go#L762)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Looks like you haven't pushed that TODO. 
</blockquote></details>

Added a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-KpHNngubhWOpDgj1iuW:boopowq)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_command.go, line 1893 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KpHfF9Q5blQpVUqm5gU:-KpHfF9Q5blQpVUqm5gV:bko6a1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/499c50bcc6d515f67a12ce7979c4d255c1b21a5d/pkg/storage/replica_command.go#L1893)):*
> ```Go
> 	// truncation.
> 	start := engine.MakeMVCCMetadataKey(keys.RaftLogKey(cArgs.EvalCtx.RangeID(), 0))
> 	end := engine.MakeMVCCMetadataKey(keys.RaftLogKey(cArgs.EvalCtx.RangeID(), args.Index-1).PrefixEnd())
> ```

Was the a reason you changed `end` here? I'm not seeing why it makes a difference.

---

*[pkg/storage/replica_command.go, line 1896 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KpHdTugYJ4PN9ZLaYAn:-KpHdTugYJ4PN9ZLaYAo:bym7fmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/499c50bcc6d515f67a12ce7979c4d255c1b21a5d/pkg/storage/replica_command.go#L1896)):*
> ```Go
> 
> 	var ms enginepb.MVCCStats
> 	if util.IsMigrated() {
> ```

@spencerkimball Per my comment in #16977, I think the ergonomics here are somewhat important. I'd like this to look something like a cluster setting:

```
var inconsistentRaftLogTruncation = settings.RegisterClusterVersionRequirement("1.1-alpha.20170608")
...
if inconsistentRaftLogTruncation.enabled() {
}
```

---

*[pkg/storage/replica_proposal.go, line 773 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KpHesT_3uZR2PnJirQZ:-KpHesT_3uZR2PnJirQ_:bg3afhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/499c50bcc6d515f67a12ce7979c4d255c1b21a5d/pkg/storage/replica_proposal.go#L773)):*
> ```Go
> 			} else {
> 				iter.Close()
> 			}
> ```

Closing `iter` on both code paths is a bit awkward. Could restructure as:

```
  iter := r.store.engine.NewIterator(...)
  err := ...
  iter.Close()
  if err != nil {
    log.Errorf(...)
  }
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-KpHfQY0rkbYSWFgdawO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR! Are you OK'ing the first commit (`IsMigrated()` as well)?

---

Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/replica_command.go, line 1893 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KpHfF9Q5blQpVUqm5gU:-KpHijKnhvAovUS3Hv7n:buotg9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/499c50bcc6d515f67a12ce7979c4d255c1b21a5d/pkg/storage/replica_command.go#L1893)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Was the a reason you changed `end` here? I'm not seeing why it makes a difference.
</blockquote></details>

No good reason, changing it back.

---

*[pkg/storage/replica_proposal.go, line 773 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-KpHesT_3uZR2PnJirQZ:-KpHiWFRtagtGbvhyZwa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/499c50bcc6d515f67a12ce7979c4d255c1b21a5d/pkg/storage/replica_proposal.go#L773)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Closing `iter` on both code paths is a bit awkward. Could restructure as:

```
  iter := r.store.engine.NewIterator(...)
  err := ...
  iter.Close()
  if err != nil {
    log.Errorf(...)
  }
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-KpHiXf5d1njgYPFCE-o:b-rjbwh8)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Yes, `IsMigrated` looks fine in the short term. 

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16993#-:-KpHk-yhMZxsZgw-pA32:b-eqonwh)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: don't send log truncations through Raft

Sending log truncations through Raft is inefficient: the Raft log is not itself
part of the replicated state. Instead, we only replicate the TruncatedState and,
as a side effect, ClearRange() the affected key range.

This is an individual performance optimization whose impact we should measure;
anecdotally it always looked like we were doing a lot of work for truncations
during a write-heavy workload; this should alleviate this somewhat). As
explained above, the change isn't made for performance at this point, though.

It also removes one migration concern for #16809, see
https://github.com/cockroachdb/cockroach/pull/16809#issuecomment-314533973.

We'll need to migrate this. It's straightforward with the in-flight PR #16977.

- we're moving logic downstream of Raft. However, we can easily migrate it
  upstream again, without a real migration, though I don't think that's going
  to happen.
- the big upshot is hopefully a large reduction in complexity for
  @irfansharif's PR: log truncation is one of the odd cases that requires
  a RaftWriteBatch. #16749 is the only other one, and there the (correct)
  solution also involves going downstream of Raft for a Raft-related write. So,
  after solving both of those, I think RaftWriteBatch can go? cc @irfansharif
- as @petermattis pointed out, after @irfansharif's change, we should be able
  to not sync the base engine on truncation changes but do it only as we
  actually clear the log entries (which can be delayed as we see fit). So for
  1000 log truncations across many ranges, we'll only have to sync once if
  that's how we set it up.
